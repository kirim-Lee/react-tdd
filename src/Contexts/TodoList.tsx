import { createContext, useEffect, useState } from 'react';

interface IContext {
  readonly todoList: string[];
  readonly addTodo: (toDo: string) => void;
  readonly deleteTodo: (index: number) => void;
}

const TodoListContext = createContext<IContext>({
  todoList: [],
  addTodo: () => {},
  deleteTodo: () => {},
});

interface IProps {
  children: JSX.Element | JSX.Element[];
}

const TODO_LIST = 'TodoList';

const TodoListProvider = ({ children }: IProps): JSX.Element => {
  const [todoList, setTodoList] = useState<string[]>([]);

  const addTodo = (toDo: string) => {
    if (!toDo) return;

    const newList = [...todoList, toDo];
    localStorage.setItem(TODO_LIST, JSON.stringify(newList));
    setTodoList(newList);
  };

  const deleteTodo = (index: number) => {
    let list = [...todoList];
    list.splice(index, 1);
    localStorage.setItem(TODO_LIST, JSON.stringify(list));
    setTodoList(list);
  };

  useEffect(() => {
    const list = localStorage.getItem(TODO_LIST);
    if (list) {
      setTodoList(JSON.parse(list));
    }
  }, []);

  return (
    <TodoListContext.Provider value={{ todoList, addTodo, deleteTodo }}>
      {children}
    </TodoListContext.Provider>
  );
};

export { TodoListContext, TodoListProvider };
