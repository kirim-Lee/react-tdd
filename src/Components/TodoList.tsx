import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import { TodoItem } from './TodoItem';
import { TodoListContext } from '../Contexts/TodoList';

const Container = styled.div`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
`;

export const TodoList = () => {
  const { todoList, deleteTodo } = useContext(TodoListContext);

  return (
    <Container data-testid="toDoList">
      {todoList.map((item, index) => (
        <TodoItem
          key={item}
          id={index}
          label={item}
          onDelete={() => deleteTodo(index)}
        />
      ))}
    </Container>
  );
};
