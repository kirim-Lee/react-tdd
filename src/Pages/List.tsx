import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Container, TodoList } from '../Components';

const AddButton = styled(Link)`
  font-size: 20px;
  color: #fff;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 60px;
  height: 60px;
  border-radius: 30px;
  cursor: pointer;
  position: absolute;
  bottom: -30px;
  background-color: #304ffe;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
  text-decoration: none;
  &:hover {
    background-color: #1e40ff;
  }
  &:active {
    box-shadow: inset 5px 5px 10px rgba(0, 0, 0, 0.2);
  }
`;

const List = () => {
  return (
    <Container>
      <TodoList />
      <AddButton to="/add">+</AddButton>
    </Container>
  );
};

export default List;
