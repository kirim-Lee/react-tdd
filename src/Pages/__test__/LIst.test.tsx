import { fireEvent, render, screen } from '@testing-library/react';
import List from '../List';
import { TodoListProvider } from '../../Contexts/TodoList';
import {
  RouterProvider,
  createMemoryRouter,
  useLocation,
} from 'react-router-dom';

const TODO_LIST = 'TodoList';

describe('List Page', () => {
  it('렌더링 테스트', () => {
    const router = createMemoryRouter([{ path: '*', element: <List /> }]);

    localStorage.setItem(
      TODO_LIST,
      JSON.stringify(['Todo 1', 'Todo 2', 'Todo 3'])
    );

    const { container } = render(
      <TodoListProvider>
        <RouterProvider router={router} />
      </TodoListProvider>
    );

    const todoItem1 = screen.getByText('Todo 1');
    expect(todoItem1).toBeInTheDocument();
    expect(todoItem1.getAttribute('href')).toBe('/detail/0');

    const todoItem2 = screen.getByText('Todo 2');
    expect(todoItem2).toBeInTheDocument();
    expect(todoItem2.getAttribute('href')).toBe('/detail/1');

    const todoItem3 = screen.getByText('Todo 3');
    expect(todoItem3).toBeInTheDocument();
    expect(todoItem3.getAttribute('href')).toBe('/detail/2');

    expect(screen.getAllByText('삭제').length).toBe(3);

    const addButton = screen.getByText('+');
    expect(addButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('할일 삭제', () => {
    const router = createMemoryRouter([{ path: '*', element: <List /> }]);

    localStorage.setItem(
      TODO_LIST,
      JSON.stringify(['Todo 1', 'Todo 2', 'Todo 3'])
    );

    render(
      <TodoListProvider>
        <RouterProvider router={router} />
      </TodoListProvider>
    );

    const todoItem = screen.getByText('Todo 2');
    expect(todoItem).toBeInTheDocument();
    fireEvent.click(todoItem.nextElementSibling as HTMLElement);
    expect(todoItem).not.toBeInTheDocument();
    expect(JSON.parse(localStorage.getItem(TODO_LIST))).not.toContain('Todo 2');
  });

  it('상세 페이지로 이동', () => {
    const TestComponent = () => {
      const { pathname } = useLocation();
      return (
        <>
          <div>{pathname}</div>
          <List />
        </>
      );
    };

    const router = createMemoryRouter([
      { path: '*', element: <TestComponent /> },
    ]);

    localStorage.setItem(
      TODO_LIST,
      JSON.stringify(['Todo 1', 'Todo 2', 'Todo 3'])
    );

    render(
      <TodoListProvider>
        <RouterProvider router={router} />
      </TodoListProvider>
    );

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const todoItem1 = screen.getByText('Todo 2');
    expect(todoItem1.getAttribute('href')).toBe('/detail/1');
    fireEvent.click(todoItem1);

    expect(url.textContent).toBe('/detail/1');
  });

  it('생성 페이지로 이동', () => {
    const TestComponent = () => {
      const { pathname } = useLocation();
      return (
        <>
          <div>{pathname}</div>
          <List />
        </>
      );
    };

    const router = createMemoryRouter([
      { path: '*', element: <TestComponent /> },
    ]);

    localStorage.setItem(
      TODO_LIST,
      JSON.stringify(['Todo 1', 'Todo 2', 'Todo 3'])
    );

    render(
      <TodoListProvider>
        <RouterProvider router={router} />
      </TodoListProvider>
    );

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const addButton = screen.getByText('+');
    fireEvent.click(addButton);

    expect(url.textContent).toBe('/add');
  });
});
