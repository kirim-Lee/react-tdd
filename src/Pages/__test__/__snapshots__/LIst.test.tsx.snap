// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Page 렌더링 테스트 1`] = `
<div>
  <div
    class="sc-bcXHqe jnXeEi"
  >
    <div
      class="sc-jrcTuL hviBsy"
      data-testid="toDoList"
    >
      <div
        class="sc-iBYQkv hdVmex"
      >
        <a
          class="sc-ftTHYK deRNsE"
          href="/detail/0"
        >
          Todo 1
        </a>
        <button
          class="sc-eDvSVe gBjEPJ"
        >
          <div
            class="sc-jSUZER jTTUoN"
          >
            삭제
          </div>
        </button>
      </div>
      <div
        class="sc-iBYQkv hdVmex"
      >
        <a
          class="sc-ftTHYK deRNsE"
          href="/detail/1"
        >
          Todo 2
        </a>
        <button
          class="sc-eDvSVe gBjEPJ"
        >
          <div
            class="sc-jSUZER jTTUoN"
          >
            삭제
          </div>
        </button>
      </div>
      <div
        class="sc-iBYQkv hdVmex"
      >
        <a
          class="sc-ftTHYK deRNsE"
          href="/detail/2"
        >
          Todo 3
        </a>
        <button
          class="sc-eDvSVe gBjEPJ"
        >
          <div
            class="sc-jSUZER jTTUoN"
          >
            삭제
          </div>
        </button>
      </div>
    </div>
    <a
      class="sc-kDvujY kxypwL"
      href="/add"
    >
      +
    </a>
  </div>
</div>
`;
